crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,301 @@
+//! - Re-send requested ([OpId], [AgentId]) to the internal channel again. It will be removed
+//!   from the list of ops to fetch if it is received in the meantime, and thus prevent a redundant
+//!   fetch request.
+//!
+//! ### Incoming op task
+//!
+//! - Incoming op is written to the data store.
+//! - Once persisted successfully, op is removed from the data object.
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+    time::Instant,
+};
+
+use kitsune2_api::{
+    builder,
+    config::ModConfig,
+    fetch::{DynFetch, DynFetchFactory, Fetch, FetchFactory},
+    tx::Transport,
+    AgentId, BoxFut, K2Result, OpId,
+};
+use tokio::sync::{
+    mpsc::{channel, Receiver, Sender},
+    Mutex,
+};
+
+const MOD_NAME: &str = "Fetch";
+
+/// Configuration parameters for [CoreFetchFactory].
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoreFetchConfig {
+    /// How many parallel op fetch requests can be made at once. Default: 2.  
+    pub parallel_request_count: u8,
+    /// Duration in ms to keep an unresponsive agent on the cool-down list. Default: 10_000.
+    pub cool_down_interval_ms: u64,
+}
+
+impl Default for CoreFetchConfig {
+    fn default() -> Self {
+        Self {
+            parallel_request_count: 2,
+            cool_down_interval_ms: 10_000,
+        }
+    }
+}
+
+impl ModConfig for CoreFetchConfig {}
+
+// TODO: Temporary trait object of a transport module to facilitate unit tests.
+type DynTransport = Arc<Mutex<dyn Transport>>;
+
+#[derive(Debug)]
+struct CoreFetch(Inner);
+
+impl CoreFetch {
+    fn new(config: CoreFetchConfig, transport: DynTransport) -> Self {
+        let inner = Inner::new();
+        inner.create_fetch_queue(config, transport);
+        Self(inner)
+    }
+}
+
+impl Fetch for CoreFetch {
+    fn add_ops(
+        &mut self,
+        op_list: Vec<OpId>,
+        source: AgentId,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            // Add ops to set.
+            let ops = &mut self.0.state.lock().await.ops;
+            op_list.clone().into_iter().for_each(|op_id| {
+                ops.insert((op_id, source.clone()));
+            });
+
+            // Pass ops to fetch tasks.
+            let futures = op_list.into_iter().map(|op_id| async {
+                if let Err(err) =
+                    self.0.fetch_queue_tx.send((op_id, source.clone())).await
+                {
+                    eprintln!(
+                        "could not pass fetch request to fetch task: {err}"
+                    );
+                }
+            });
+            futures::future::join_all(futures).await;
+
+            Ok(())
+        })
+    }
+}
+
+type FetchRequest = (OpId, AgentId);
+
+#[derive(Debug)]
+struct State {
+    ops: HashSet<FetchRequest>,
+    cool_down_list: HashMap<AgentId, Instant>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    state: Arc<Mutex<State>>,
```

Comment:
One mutex for all state objects.

Reply:
This needs to not be a tokio Mutex. Please switch to a std::sync::Mutex.

Reply:
Switched https://github.com/holochain/kitsune2/pull/35/commits/9add269447278ecaf2a1e4a4fad8142cf8c61011

Must the receiver use a tokio mutex?

Reply:
Correct, when you need to await anything while a mutex is locked you need it to be a tokio mutex. But that makes them much slower system-wise, so should only be used in very calculated situations.